src/backend/distributed/ddl/foreign_constraint.c

```diff
@@ -11,21 +11,492 @@
+												   HASH_ENTER, &referencingFound);
+		referencedNode = (FRelNode *) hash_search(frelGraph->nodeMap,
+												  &referencedRelationOid,
+												  HASH_ENTER, &referencedFound);
+
+		/*
+		 * If we found a node in the graph we only need to add referenced node to
+		 * the adjacency  list of that node.
+		 */
+		if (referencingFound)
+		{
+			/*
+			 * If referenced node is already in the adjacency list of referenced node, do nothing.
+			 */
+			if (referencingNode->adjacencyList == NIL || !list_member(
+					referencingNode->adjacencyList, referencedNode))
+			{
+				/*
+				 * If referenced node also exists, add it to the adjacency list
+				 * and continue.
+				 */
+				if (referencedFound)
+				{
+					referencingNode->adjacencyList = lappend(
+						referencingNode->adjacencyList, referencedNode);
+					referencedNode->backAdjacencyList = lappend(
+						referencedNode->backAdjacencyList, referencingNode);
+				}
+				else
+				{
+					/*
+					 * We need to setup the node and add it to the adjacency list
+					 * of referencing node.
+					 */
+					referencedNode->adjacencyList = NIL;
+					referencedNode->backAdjacencyList = NIL;
+					referencedNode->seen = false;
+					referencingNode->adjacencyList = lappend(
+						referencingNode->adjacencyList, referencedNode);
+					referencedNode->backAdjacencyList = lappend(
+						referencedNode->backAdjacencyList, referencingNode);
+				}
+			}
+			else
+			{
+				continue;
+			}
+		}
+		else
+		{
+			/*
+			 * If referencing node is not exist in the graph, set its remaining parameters.
+			 */
+			referencingNode->adjacencyList = NIL;
+			referencingNode->backAdjacencyList = NIL;
+			referencingNode->seen = false;
+
+			if (referencedFound)
+			{
+				referencingNode->adjacencyList = lappend(referencingNode->adjacencyList,
+														 referencedNode);
+				referencedNode->backAdjacencyList = lappend(
+					referencedNode->backAdjacencyList, referencingNode);
+			}
+			else
+			{
+				/*
+				 * We need to setup the node and add it to the adjacency list
+				 * of referencing node.
+				 */
+				referencedNode->adjacencyList = NIL;
+				referencedNode->backAdjacencyList = NIL;
+				referencedNode->seen = false;
+				referencingNode->adjacencyList = lappend(referencingNode->adjacencyList,
+														 referencedNode);
+				referencedNode->backAdjacencyList = lappend(
+					referencedNode->backAdjacencyList, referencingNode);
+			}
+		}
+	}
+
+	frelGraph->isValid = true;
+	systable_endscan(fkeyScan);
+	heap_close(fkeyRel, AccessShareLock);
+	MemoryContextSwitchTo(oldContext);
+}
+
+
+/*
+ * ClearForeignKeyRelationGraph clear all the allocated memory obtained for
+ * foreign key relation graph. Since all the variables of relation graph
+ * obtained within the same context, destroying hash map is enough as it
+ * deletes the context.
+ */
+void
+ClearForeignKeyRelationGraphContext()
+{
+	if (frelGraph == NULL)
+	{
+		return;
+	}
+
+	hash_destroy(frelGraph->nodeMap);
```

Comment:
Since we allocate `frelGraph ` under the same memory context, let's set `frelGraph = NULL` after `hash_destroy` given that all the graph will be already freed.